misc/config_tools/launch_config/launch_cfg_gen.py

```diff
@@ -1,269 +1,366 @@
+        params = virtio_network.split(",", maxsplit=1)
+        tap_conf = f"tap={params[0]}"
+        params = [tap_conf] + params[1:]
+        script.add_init_command(f"mac=$(cat /sys/class/net/e*/address)")
+        params.append(f"mac_seed=${{mac:0:17}}-{vm_name}")
+        script.add_virtual_device("virtio-net", options=",".join(params))
+
+    for virtio_block in vm_launch_etree.xpath(".//virtio_devices/block[text() != '']/text()"):
+        params = virtio_block.split(":", maxsplit=1)
+        if len(params) == 1:
+            script.add_virtual_device("virtio-blk", options=virtio_block)
+        else:
+            block_device = params[0]
+            rootfs_img = params[1]
+            var = f"dir_{os.path.basename(block_device)}"
+            script.add_init_command(f"{var}=`mount_partition {block_device}`")
+            script.add_virtual_device("virtio-blk", options=os.path.join(f"${{{var}}}", rootfs_img))
+            script.add_deinit_command(f"unmount_partition ${{{var}}}")
+
+    # Passthrough PCI devices
+    bdf_regex = re.compile("([0-9a-f]{2}):([0-1][0-9a-f]).([0-7])")
+    for passthru_device in vm_launch_etree.xpath(".//passthrough_devices/*/text()"):
+        m = bdf_regex.match(passthru_device)
+        if not m:
+            continue
+        script.add_passthru_device(int(m.group(1), 16), int(m.group(2), 16), int(m.group(3), 16))
+
+    for sriov_gpu_device in vm_launch_etree.xpath(".//sriov/gpu/text()"):
+        m = bdf_regex.match(sriov_gpu_device)
+        if not m:
+            continue
+        script.add_passthru_device(int(m.group(1), 16), int(m.group(2), 16), int(m.group(3), 16), options="igd-vf")
+
+    ###
+    # Miscellaneous
+    ###
+    script.add_dynamic_dm_parameter("add_rtvm_options")
+    script.add_dynamic_dm_parameter("add_logger_settings", "console=4 kmsg=3 disk=5")
+
+    ###
+    # Lastly, conclude the device model parameters with the VM name
+    ###
+    script.add_plain_dm_parameter(f"{vm_name}")
+
+    return script
+
+def main(board_xml, scenario_xml, launch_xml, user_vmid, out_dir):
+    board_etree = etree.parse(board_xml)
+    scenario_etree = etree.parse(scenario_xml)
+    launch_etree = etree.parse(launch_xml)
+
+    sos_vm_id = eval_xpath(scenario_etree, "//vm[vm_type='SERVICE_VM']/@id")
+    post_vms = scenario_etree.xpath("//vm[starts-with(vm_type, 'POST_')]")
+    if sos_vm_id is None and len(post_vms) > 0:
+        logging.error("The scenario does not define a service VM so no launch scripts will be generated for the post-launched VMs in the scenario.")
+        return 1
+    sos_vm_id = int(sos_vm_id)
+
+    try:
+        os.mkdir(out_dir)
+    except FileExistsError:
+        if os.path.isfile(out_dir):
+            logging.error(f"Cannot create output directory {out_dir}: File exists")
+            return 1
+    except Exception as e:
+        logging.error(f"Cannot create output directory: {e}")
+        return 1
+
+    if user_vmid == 0:
```

Comment:
Nitpick: should we use `user_vm_id` instead to be consistent with `service_vm_id` (and `post_vm_id`, etc.)?

Reply:
I'll update this for internal variables, but keep the command line interface as `--user-vmid` for now as this is used by the premerge tests.

After we integrate this script into the build system and the premerge test scripts get rid of invoking this script directly, we can refine those interfaces.